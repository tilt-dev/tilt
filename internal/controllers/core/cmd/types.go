package cmd

import "github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"

type Cmd = v1alpha1.Cmd
type CmdSpec = v1alpha1.CmdSpec
type CmdStatus = v1alpha1.CmdStatus
type CmdStateRunning = v1alpha1.CmdStateRunning
type CmdStateTerminated = v1alpha1.CmdStateTerminated

const LabelManifest = v1alpha1.LabelManifest
