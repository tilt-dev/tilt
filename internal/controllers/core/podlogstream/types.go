package podlogstream

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type PodLogStream = v1alpha1.PodLogStream
type PodLogStreamSpec = v1alpha1.PodLogStreamSpec
type PodLogStreamStatus = v1alpha1.PodLogStreamStatus
type ContainerLogStreamStatus = v1alpha1.ContainerLogStreamStatus
type ObjectMeta = metav1.ObjectMeta
