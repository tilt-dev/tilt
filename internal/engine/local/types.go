package local

import (
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/tilt-dev/tilt/pkg/apis/core/v1alpha1"
)

type Cmd = v1alpha1.Cmd
type CmdList = v1alpha1.CmdList
type CmdStatus = v1alpha1.CmdStatus
type CmdSpec = v1alpha1.CmdSpec
type CmdStateWaiting = v1alpha1.CmdStateWaiting
type CmdStateTerminated = v1alpha1.CmdStateTerminated
type CmdStateRunning = v1alpha1.CmdStateRunning
type ObjectMeta = metav1.ObjectMeta
