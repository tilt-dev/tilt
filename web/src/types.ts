import { Alert } from "./alerts"
import { Facet } from "./facets"

export enum SocketState {
  Loading,
  Reconnecting,
  Closed,
  Active,
}

export enum ResourceView {
  Log,
  Alerts,
  Facets = 2,
}

export enum TriggerMode {
  TriggerModeAuto,
  TriggerModeManualAfterInitial,
  TriggerModeManualIncludingInitial,
}

// what is the status of the resource in the cluster
export enum RuntimeStatus {
  Ok = "ok",
  Pending = "pending",
  Error = "error",
  NotApplicable = "not_applicable",
}

// What is the status of the resource with respect to Tilt
export enum ResourceStatus {
  Building, // Tilt is actively doing something (e.g., docker build or kubectl apply)
  Pending, // not building, healthy, or unhealthy, but presumably on its way to one of those (e.g., queued to build, or ContainerCreating)
  Healthy, // e.g., build succeeded and pod is running and healthy
  Unhealthy, // e.g., last build failed, or CrashLoopBackOff
  Warning, // e.g., an undismissed restart
  None, // e.g., a manual build that has never executed
}

export type SnapshotHighlight = {
  beginningLogID: string
  endingLogID: string
  text: string
}

export enum ShowFatalErrorModal {
  Default,
  Show,
  Hide,
}

export type Snapshot = {
  view: Proto.webviewView
  isSidebarClosed: boolean
  path?: string
  snapshotHighlight?: SnapshotHighlight | null
}

// A plaintext representation of a line of the log,
// with metadata to render it in isolation.
//
// The metadata should be stored as primitive fields
// so that React's default caching behavior will kick in.
export type LogLine = {
  // We assume that 'text' does not contain a newline
  text: string
  manifestName: string
  level: string
}
