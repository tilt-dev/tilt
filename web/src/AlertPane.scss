@import "constants";

.AlertPane {
  padding: $spacing-unit / 2;
  width: 100%;
  display: flex;
  justify-content: center;
  box-sizing: border-box;
}

.AlertPane ul {
  list-style: none;
  width: 100%;
}
.AlertPane-item {
  position: relative;
}
.AlertPane-item > header {
  background-color: $color-gray;
  border: 1px solid $color-gray-light;
  display: flex;
  flex-direction: column;
  padding-top: $spacing-unit / 4;
  padding-bottom: $spacing-unit / 4;
  padding-left: $spacing-unit / 2;
  padding-right: $spacing-unit / 2;
}

.AlertPane-headerDiv {
  display: flex;
  justify-content: space-between;
}

.AlertPane-headerDiv span + span {
  margin-left: $spacing-unit / 2;
  border-left: 1px solid $color-gray-light;
  padding-left: $spacing-unit / 2;
}

.AlertPane-headerDiv-header {
  margin: 0;
  font-size: $font-size;
}

.AlertPane-item > section {
  background-color: $color-gray-dark;
  border-bottom-left-radius: $spacing-unit / 4;
  border-bottom-right-radius: $spacing-unit / 4;
  border-left: 1px solid $color-gray-light;
  border-right: 1px solid $color-gray-light;
  border-bottom: 1px solid $color-gray-light;
  padding: $spacing-unit / 2;
  min-height: 2em;
}

.AlertPane-item + .AlertPane-item {
  margin-top: $spacing-unit;
}

// Hide all lines by default
.AlertPane-item code {
  display: none;
  white-space: pre-wrap;
}

// ...only show last few lines
.AlertPane-item code:nth-last-child(-n + 8) {
  display: inline;
  min-height: $spacing-unit / 2; // Respect blank lines
  height: auto;
  opacity: 1;
}

.AlertPane-dismissButton {
  position: absolute;
  right: $spacing-unit/2;
  bottom: $spacing-unit/2;
  align-content: center;
  background-color: $color-gray-dark;
  color: $color-blue;
  @include button-text;
  border: 1px solid $color-blue;
  border-radius: 5px;
  padding: $spacing-unit/3;
  min-width: $spacing-unit * 3;
  transition: background-color ease 0.3s;
}

.AlertPane-dismissButton:hover {
  background-color: $color-gray;
  cursor: pointer;
}