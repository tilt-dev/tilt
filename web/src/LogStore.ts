// Client-side log store, which helps client side rendering and filtering of logs.
//
// Loosely adapted from the data structures in
// pkg/model/logstore/logstore.go
// but with better support for incremental updates and rendering.

import { LogLine } from "./types"

// Firestore doesn't properly handle maps with keys equal to the empty string, so
// we normalize all empty span ids to '_' client-side.
const defaultSpanId = "_"

type LogSpan = {
  manifestName: string
  lastSegmentIndex: number
  firstSegmentIndex: number
}

class LogSegment {
  spanId: string
  time: string
  text: string
  level: string
  anchor: boolean
  continuesLine: boolean

  constructor(seg: Proto.webviewLogSegment) {
    this.spanId = seg.spanId || defaultSpanId
    this.time = seg.time ?? ""
    this.text = seg.text ?? ""
    this.level = seg.level ?? "INFO"
    this.anchor = seg.anchor ?? false
    this.continuesLine = false
  }

  startsLine() {
    return !this.continuesLine
  }

  isComplete() {
    return this.text[this.text.length - 1] == "\n"
  }

  canContinueLine(seg: LogSegment) {
    return this.level == seg.level && this.spanId == seg.spanId
  }
}

type LogWarning = {
  anchorIndex: number
  spanId: string
  text: string
}

class LogStore {
  spans: { [key: string]: LogSpan }
  segments: LogSegment[]
  checkpoint: number

  // We index all the warnings up-front by span id.
  warningIndex: { [key: string]: LogWarning[] }

  constructor() {
    this.spans = {}
    this.segments = []
    this.checkpoint = 0
    this.warningIndex = {}
  }

  warnings(spanId: string): LogWarning[] {
    return this.warningIndex[spanId] ?? []
  }

  toLogList(): Proto.webviewLogList {
    let spans = {} as { [key: string]: Proto.webviewLogSpan }
    for (let key in this.spans) {
      spans[key] = { manifestName: this.spans[key].manifestName }
    }

    let segments = this.segments.map(
      (segment): Proto.webviewLogSegment => {
        let spanId = segment.spanId
        let time = segment.time
        let text = segment.text
        let level = segment.level
        return { spanId, time, text, level }
      }
    )
    return {
      spans: spans,
      segments: segments,
    }
  }

  append(logList: Proto.webviewLogList) {
    let newSpans = logList.spans as { [key: string]: Proto.webviewLogSpan }
    let newSegments = logList.segments ?? []
    let fromCheckpoint = logList.fromCheckpoint ?? 0
    let toCheckpoint = logList.toCheckpoint ?? 0
    if (fromCheckpoint < 0) {
      return
    }

    if (fromCheckpoint < this.checkpoint) {
      // The server is re-sending some logs we already have, so slice them off.
      let deleteCount = this.checkpoint - fromCheckpoint
      newSegments = newSegments.slice(deleteCount)
    }

    if (toCheckpoint > this.checkpoint) {
      this.checkpoint = toCheckpoint
    }

    for (let key in newSpans) {
      let spanId = key || defaultSpanId
      let exists = this.spans[spanId]
      if (!exists) {
        this.spans[spanId] = {
          manifestName: newSpans[key].manifestName ?? "",
          firstSegmentIndex: -1,
          lastSegmentIndex: -1,
        }
      }
    }

    newSegments.forEach(segment => {
      let newSegment = new LogSegment(segment)
      let span = this.spans[newSegment.spanId]
      if (!span) {
        // If we don't have the span for this log, we can't meaningfully print it,
        // so just drop it. This means that there's a bug on the server, and
        // the best the client can do is fail gracefully.
        return
      }
      let isStartingNewLine = false
      if (span.lastSegmentIndex == -1) {
        isStartingNewLine = true
      } else {
        let seg = this.segments[span.lastSegmentIndex]
        if (seg.isComplete() || !seg.canContinueLine(newSegment)) {
          isStartingNewLine = true
        }
      }

      if (span.firstSegmentIndex == -1) {
        span.firstSegmentIndex = this.segments.length
      }
      span.lastSegmentIndex = this.segments.length

      newSegment.continuesLine = !isStartingNewLine
      this.segments.push(newSegment)
    })
  }

  allLog(): LogLine[] {
    return this.logHelper(this.spans)
  }

  spanLog(spanIds: string[]): LogLine[] {
    let spans: { [key: string]: LogSpan } = {}
    spanIds.forEach(spanId => {
      spanId = spanId ? spanId : defaultSpanId
      let span = this.spans[spanId]
      if (span) {
        spans[spanId] = span
      }
    })

    return this.logHelper(spans)
  }

  spansForManifest(mn: string): { [key: string]: LogSpan } {
    let result: { [key: string]: LogSpan } = {}
    for (let spanId in this.spans) {
      let span = this.spans[spanId]
      if (span.manifestName == mn) {
        result[spanId] = span
      }
    }
    return result
  }

  manifestLog(mn: string): LogLine[] {
    let spans = this.spansForManifest(mn)
    return this.logHelper(spans)
  }

  logHelper(spansToLog: { [key: string]: LogSpan }): LogLine[] {
    let result: LogLine[] = []
    let lastLineCompleted = false

    // We want to print the log line-by-line, but we don't actually store the logs
    // line-by-line. We store them as segments.
    //
    // This means we need to:
    // 1) At segment x,
    // 2) If x starts a new line, print it, then run ahead to print the rest of the line
    //    until the entire line is consumed.
    // 3) If x does not start a new line, skip it, because we assume it was handled
    //    in a previous line.
    //
    // This can have some O(n^2) perf characteristics in the worst case, but
    // for normal inputs should be fine.
    let startIndex = 0
    let lastIndex = this.segments.length - 1
    let isFilteredLog =
      Object.keys(spansToLog).length != Object.keys(this.spans).length
    if (isFilteredLog) {
      let earliestStartIndex = -1
      let latestEndIndex = -1
      for (let spanId in spansToLog) {
        let span = spansToLog[spanId]
        if (
          earliestStartIndex == -1 ||
          span.firstSegmentIndex < earliestStartIndex
        ) {
          earliestStartIndex = span.firstSegmentIndex
        }
        if (latestEndIndex == -1 || span.lastSegmentIndex > latestEndIndex) {
          latestEndIndex = span.lastSegmentIndex
        }
      }

      if (earliestStartIndex == -1) {
        return []
      }

      startIndex = earliestStartIndex
      lastIndex = latestEndIndex
    }

    let isFirstLine = true
    let currentLine = {}
    for (let i = startIndex; i <= lastIndex; i++) {
      let segment = this.segments[i]
      if (!segment.startsLine()) {
        continue
      }

      let spanId = segment.spanId
      let span = spansToLog[spanId]
      if (!span) {
        continue
      }

      let currentLine = {
        manifestName: span.manifestName,
        text: segment.text,
        level: segment.level,
      }
      isFirstLine = false

      // If this segment is not complete, run ahead and try to complete it.
      if (segment.isComplete()) {
        lastLineCompleted = true

        // strip off the newline
        currentLine.text = currentLine.text.substring(
          0,
          currentLine.text.length - 1
        )
        result.push(currentLine)
        continue
      }

      lastLineCompleted = false
      for (
        let currentIndex = i + 1;
        currentIndex <= span.lastSegmentIndex;
        currentIndex++
      ) {
        let currentSeg = this.segments[currentIndex]
        if (currentSeg.spanId != spanId) {
          continue
        }

        if (!currentSeg.canContinueLine(segment)) {
          break
        }

        currentLine.text += currentSeg.text
        if (currentSeg.isComplete()) {
          lastLineCompleted = true

          // strip off the newline
          currentLine.text = currentLine.text.substring(
            0,
            currentLine.text.length - 1
          )
          break
        }
      }
      result.push(currentLine)
    }

    return result
  }
}

export default LogStore
