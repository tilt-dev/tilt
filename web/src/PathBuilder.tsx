// A little helper class for building paths relative to the root of the app.

class PathBuilder {
  private host: string
  private snapId: string = ""

  constructor(host: string, pathname: string) {
    this.host = host

    const snapshotRe = new RegExp("^/snapshot/([^/]+)")
    let snapMatch = snapshotRe.exec(pathname)
    if (snapMatch) {
      this.snapId = snapMatch[1]
    }
  }

  getDataUrl() {
    if (this.isSnapshot()) {
      return this.snapshotDataUrl()
    }
    return `ws://${this.host}/ws/view`
  }

  isSnapshot(): boolean {
    return this.snapId !== ""
  }

  private snapshotDataUrl(): string {
    return `/api/snapshot/${this.snapId}`
  }

  private snapshotPathBase(): string {
    return `/snapshot/${this.snapId}`
  }

  rootPath() {
    if (this.isSnapshot()) {
      return this.snapshotPathBase()
    }
    return ""
  }

  path(relPath: string) {
    if (relPath[0] !== "/") {
      throw new Error('relPath should start with "/", actual:' + relPath)
    }
    return this.rootPath() + relPath
  }
}

export default PathBuilder
