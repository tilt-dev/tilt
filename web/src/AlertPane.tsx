import React, { PureComponent } from "react"
import { ReactComponent as LogoWordmarkSvg } from "./assets/svg/logo-wordmark-gray.svg"
import AnsiLine from "./AnsiLine"
import TimeAgo from "react-timeago"
import "./AlertPane.scss"
import { timeAgoFormatter } from "./timeFormatters"
import { getResourceAlerts, hasAlert } from "./alerts"
import PathBuilder from "./PathBuilder"
import LogStore from "./LogStore"

type Resource = Proto.webviewResource

type AlertsProps = {
  pathBuilder: PathBuilder
  resources: Array<Resource>
  logStore: LogStore | null
}

function logToLines(s: string) {
  return s.split("\n").map((l, i) => <AnsiLine key={"logLine" + i} line={l} />)
}

class AlertPane extends PureComponent<AlertsProps> {
  render() {
    let el = (
      <section className="Pane-empty-message">
        <LogoWordmarkSvg />
        <h2>No Alerts Found</h2>
      </section>
    )

    let alerts = this.renderAlerts()
    if (alerts.length > 0) {
      el = <ul>{alerts}</ul>
    }

    return <section className="AlertPane">{el}</section>
  }

  renderAlerts() {
    let formatter = timeAgoFormatter
    let alertElements: Array<JSX.Element> = []
    let resources = this.props.resources
    let isSnapshot = this.props.pathBuilder.isSnapshot()

    let alertResources = resources.filter(r => hasAlert(r))
    alertResources.forEach(resource => {
      let resName = resource.name ?? ""
      getResourceAlerts(resource, this.props.logStore).forEach(alert => {
        let dismissButton = <div />
        if (alert.dismissHandler && !isSnapshot) {
          dismissButton = (
            <button
              className="AlertPane-dismissButton"
              onClick={alert.dismissHandler}
            >
              Dismiss
            </button>
          )
        }
        alertElements.push(
          <li key={alert.alertType + resName} className="AlertPane-item">
            <header>
              <div className="AlertPane-headerDiv">
                <h3 className="AlertPane-headerDiv-header">{alert.header}</h3>
              </div>
              <div className="AlertPane-headerDiv">
                <p>
                  <span>Resource: {alert.resourceName}</span>
                  <span>Type: {alert.alertType}</span>
                </p>
                <TimeAgo date={alert.timestamp} formatter={formatter} />
              </div>
            </header>
            <section>{logToLines(alert.msg)}</section>
            {dismissButton}
          </li>
        )
      })
    })
    return alertElements
  }
}

export default AlertPane
