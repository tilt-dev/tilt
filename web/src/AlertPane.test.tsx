import React from "react"
import AlertPane from "./AlertPane"
import renderer from "react-test-renderer"
import { oneResourceUnrecognizedError } from "./testdata"
import { TriggerMode } from "./types"
import PathBuilder from "./PathBuilder"
import { mount } from "enzyme"
import LogStore from "./LogStore"

type Resource = Proto.webviewResource

let logStore = new LogStore()

let pb = new PathBuilder("localhost", "")
beforeEach(() => {
  fetchMock.resetMocks()
  Date.now = jest.fn(() => 1482363367071)
  logStore = new LogStore()
})

it("renders no errors", () => {
  let resource = fillResourceFields()
  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()

  expect(tree).toMatchSnapshot()
})

it("renders one container start error", () => {
  const ts = "1,555,970,585,039"

  let resource = fillResourceFields()
  resource.crashLog = "Eeeeek there is a problem"
  resource.buildHistory = [
    {
      finishTime: ts,
    },
  ]
  if (!resource.k8sResourceInfo) throw new Error("Missing k8s info")
  resource.k8sResourceInfo.podCreationTime = ts
  resource.k8sResourceInfo.podStatus = "Error"
  resource.k8sResourceInfo.podRestarts = 2

  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()

  // the podStatus will flap between "Error" and "CrashLoopBackOff"
  resource.k8sResourceInfo.podStatus = "CrashLoopBackOff"
  resource.k8sResourceInfo.podRestarts = 3
  const newTree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()
  expect(newTree).toMatchSnapshot()
})

it("renders pod restart dismiss button", () => {
  let resource = fillResourceFields()
  const ts = "1,555,970,585,039"
  resource.crashLog = "Eeeeek there is a problem"
  let rInfo = resource.k8sResourceInfo
  if (!rInfo) throw new Error("Missing k8s info")
  rInfo.podName = "pod-name"
  rInfo.podCreationTime = ts
  rInfo.podStatus = "Running"
  rInfo.podRestarts = 2

  let resources: Array<Resource> = [resource]

  let root = mount(
    <AlertPane pathBuilder={pb} resources={resources} logStore={logStore} />
  )
  let button = root.find(".AlertPane-dismissButton")
  expect(button).toHaveLength(1)
  fetchMock.mockResponse(JSON.stringify({}))
  button.simulate("click")

  expect(fetchMock.mock.calls.length).toEqual(1)
  expect(fetchMock.mock.calls[0][0]).toEqual("/api/action")
  expect(fetchMock.mock.calls[0][1].body).toEqual(
    JSON.stringify({
      type: "PodResetRestarts",
      manifest_name: "foo",
      visible_restarts: 2,
      pod_id: "pod-name",
    })
  )
})

it("shows that a container has restarted", () => {
  const ts = "1,555,970,585,039"
  let resource = fillResourceFields()
  resource.crashLog = "Eeeeek the container crashed"
  resource.buildHistory = [
    {
      finishTime: ts,
    },
  ]
  if (!resource.k8sResourceInfo) throw new Error("missing k8s info")
  resource.k8sResourceInfo.podStatus = "ok"
  resource.k8sResourceInfo.podCreationTime = ts
  resource.k8sResourceInfo.podRestarts = 1
  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it("shows that a crash rebuild has occurred", () => {
  const ts = "1,555,970,585,039"
  let resource = fillResourceFields()
  resource.crashLog = "Eeeeek the container crashed"
  resource.buildHistory = [
    {
      finishTime: ts,
      isCrashRebuild: true,
    },
  ]
  if (!resource.k8sResourceInfo) throw new Error("missing k8s info")
  resource.k8sResourceInfo.podCreationTime = ts
  resource.k8sResourceInfo.podStatus = "ok"

  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it("renders multiple lines of a crash log", () => {
  const ts = "1,555,970,585,039"

  let resource = fillResourceFields()
  resource.crashLog = "Eeeeek the container crashed\nno but really it crashed"
  resource.buildHistory = [
    {
      finishTime: ts,
      isCrashRebuild: true,
    },
  ]
  if (!resource.k8sResourceInfo) throw new Error("missing k8s info")
  resource.k8sResourceInfo.podCreationTime = ts
  resource.k8sResourceInfo.podStatus = "ok"

  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it("renders warnings", () => {
  const ts = "1,555,970,585,039"
  let resource = fillResourceFields()
  resource.crashLog = "Eeeeek the container crashed"
  resource.buildHistory = [
    {
      finishTime: ts,
      isCrashRebuild: true,
      warnings: ["Hi I'm a warning"],
    },
  ]
  if (!resource.k8sResourceInfo) throw new Error("missing k8s info")
  resource.k8sResourceInfo.podCreationTime = ts
  resource.k8sResourceInfo.podStatus = "ok"

  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane
        pathBuilder={pb}
        resources={resources as Array<Resource>}
        logStore={logStore}
      />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it("renders one container unrecognized error", () => {
  const ts = "1,555,970,585,039"
  let resource = oneResourceUnrecognizedError()

  let resources = [resource]

  const tree = renderer
    .create(
      <AlertPane pathBuilder={pb} resources={resources} logStore={logStore} />
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

//TODO TFT: Create tests testing that button appears and URL appears
function fillResourceFields(): Resource {
  return {
    name: "foo",
    buildHistory: [],
    crashLog: "",
    directoriesWatched: [],
    endpoints: [],
    podID: "",
    isTiltfile: false,
    lastDeployTime: "",
    pathsWatched: [],
    pendingBuildEdits: [],
    pendingBuildReason: 0,
    pendingBuildSince: "",
    k8sResourceInfo: {
      podName: "",
      podCreationTime: "",
      podUpdateStartTime: "",
      podStatus: "",
      podStatusMessage: "",
      podRestarts: 0,
    },
    runtimeStatus: "",
    triggerMode: TriggerMode.TriggerModeAuto,
    hasPendingChanges: true,
    facets: [],
    queued: false,
  }
}
