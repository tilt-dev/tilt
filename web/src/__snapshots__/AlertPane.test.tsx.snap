// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders multiple lines of a crash log 1`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            Pod crashed
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              ResourceCrashRebuild
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed

          </span>
        </code>
        <code>
          <span
            className={null}
            style={null}
          >
            no but really it crashed

          </span>
        </code>
      </section>
      <div />
    </li>
  </ul>
</section>
`;

exports[`renders no errors 1`] = `
<section
  className="AlertPane"
>
  <section
    className="Pane-empty-message"
  >
    <svg>
      logo-wordmark-gray.svg
    </svg>
    <h2>
      No Alerts Found
    </h2>
  </section>
</section>
`;

exports[`renders one container start error 1`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              PodStatusError
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek there is a problem

          </span>
        </code>
      </section>
      <div />
    </li>
  </ul>
</section>
`;

exports[`renders one container start error 2`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              PodStatusError
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek there is a problem

          </span>
        </code>
      </section>
      <div />
    </li>
  </ul>
</section>
`;

exports[`renders one container unrecognized error 1`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              vigoda
            </span>
            <span>
              Type: 
              PodStatusError
            </span>
          </p>
          <time
            dateTime="2016-12-21T23:36:07.071Z"
            title="2016-12-21 23:36"
          >
            &lt;5s ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Detailed message on GarbleError

          </span>
        </code>
      </section>
      <div />
    </li>
  </ul>
</section>
`;

exports[`renders warnings 1`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            Pod crashed
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              ResourceCrashRebuild
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed

          </span>
        </code>
      </section>
      <div />
    </li>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            foo
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              Warning
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Hi I'm a warning

          </span>
        </code>
      </section>
      <div />
    </li>
  </ul>
</section>
`;

exports[`shows that a container has restarted 1`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            Restarts: 1
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              PodRestartError
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed

          </span>
        </code>
      </section>
      <button
        className="AlertPane-dismissButton"
        onClick={[Function]}
      >
        Dismiss
      </button>
    </li>
  </ul>
</section>
`;

exports[`shows that a crash rebuild has occurred 1`] = `
<section
  className="AlertPane"
>
  <ul>
    <li
      className="AlertPane-item"
    >
      <header>
        <div
          className="AlertPane-headerDiv"
        >
          <h3
            className="AlertPane-headerDiv-header"
          >
            Pod crashed
          </h3>
        </div>
        <div
          className="AlertPane-headerDiv"
        >
          <p>
            <span>
              Resource: 
              foo
            </span>
            <span>
              Type: 
              ResourceCrashRebuild
            </span>
          </p>
          <time
            dateTime="1949-11-18T09:39:00.000Z"
            title="1,555,970,585,039"
          >
            67yr ago
          </time>
        </div>
      </header>
      <section>
        <code>
          <span
            className={null}
            style={null}
          >
            Eeeeek the container crashed

          </span>
        </code>
      </section>
      <div />
    </li>
  </ul>
</section>
`;
